File: leetcode/78.subsets.py
# @leet start
class Solution:
    def subsets(self, nums: list[int]) -> list[list[int]]:
        res: list[list[int]] = []
        curr: list[int] = []

        def backtrack(i: int) -> None:
            if i == len(nums):
                res.append(curr[:])

                return

            curr.append(nums[i])
            backtrack(i + 1)

            curr.pop()
            backtrack(i + 1)

        backtrack(0)

        return res


# @leet end

File: leetcode/78.subsets.ts
// @leet start
function subsets(nums: number[]): number[][] {
  const result: number[][] = [];
  let curr: number[] = [];

  const backtrack = (i: number) => {
    if (i === nums.length) {
      result.push([...curr]);

      return;
    }

    backtrack(i + 1);
    curr.push(nums[i]);
    backtrack(i + 1);
    curr.pop();
  };

  backtrack(0);

  return result;
}
// @leet end
